crates/bevy_pbr/src/deferred/mod.rs

```diff
@@ -0,0 +1,451 @@
+    RenderApp,
+};
+
+use crate::{
+    EnvironmentMapLight, MeshPipelineKey, ViewFogUniformOffset, ViewLightsUniformOffset,
+    MAX_CASCADES_PER_LIGHT, MAX_DIRECTIONAL_LIGHTS,
+};
+
+pub struct PbrDeferredLightingPlugin;
+
+pub const DEFERRED_LIGHTING_SHADER_HANDLE: Handle<Shader> =
+    Handle::weak_from_u128(2708011359337029741);
+
+pub const DEFAULT_PBR_DEFERRED_LIGHTING_DEPTH_ID: u8 = 1;
+
+/// Component with a `depth_id` for specifying which corresponding materials should be rendered by this specific PBR deferred lighting pass.
+/// Will be automatically added to entities with the [`DeferredPrepass`] component that don't already have a [`PbrDeferredLightingDepthId`].
+#[derive(Component, Clone, Copy, ExtractComponent, ShaderType)]
+pub struct PbrDeferredLightingDepthId {
+    depth_id: u32,
+
+    #[cfg(all(feature = "webgl", target_arch = "wasm32"))]
+    _webgl2_padding: bevy_math::Vec3,
+}
+
+impl PbrDeferredLightingDepthId {
+    pub fn new(value: u8) -> PbrDeferredLightingDepthId {
+        PbrDeferredLightingDepthId {
+            depth_id: value as u32,
+
+            #[cfg(all(feature = "webgl", target_arch = "wasm32"))]
+            _webgl2_padding: bevy_math::Vec3::ZERO,
+        }
+    }
+
+    pub fn set(&mut self, value: u8) {
+        self.depth_id = value as u32;
+    }
+
+    pub fn get(&self) -> u8 {
+        self.depth_id as u8
+    }
+}
+
+impl Default for PbrDeferredLightingDepthId {
+    fn default() -> Self {
+        PbrDeferredLightingDepthId {
+            depth_id: 1,
+
+            #[cfg(all(feature = "webgl", target_arch = "wasm32"))]
+            _webgl2_padding: bevy_math::Vec3::ZERO,
+        }
+    }
+}
+
+impl Plugin for PbrDeferredLightingPlugin {
+    fn build(&self, app: &mut App) {
+        app.add_plugins((
+            ExtractComponentPlugin::<PbrDeferredLightingDepthId>::default(),
+            UniformComponentPlugin::<PbrDeferredLightingDepthId>::default(),
+        ))
+        .add_systems(PostUpdate, insert_deferred_lighting_depth_id_component);
+
+        load_internal_asset!(
+            app,
+            DEFERRED_LIGHTING_SHADER_HANDLE,
+            "deferred_lighting.wgsl",
+            Shader::from_wgsl
+        );
+
+        let Ok(render_app) = app.get_sub_app_mut(RenderApp) else {
+            return;
+        };
+
+        render_app
+            .init_resource::<SpecializedRenderPipelines<DeferredLightingLayout>>()
+            .add_systems(
+                Render,
+                (prepare_deferred_lighting_pipelines.in_set(RenderSet::Prepare),),
+            )
+            .add_render_graph_node::<ViewNodeRunner<DeferredLightingNode>>(
+                core_3d::graph::NAME,
+                DEFERRED_LIGHTING_PASS,
+            )
+            .add_render_graph_edges(
+                core_3d::graph::NAME,
+                &[
+                    core_3d::graph::node::START_MAIN_PASS,
+                    DEFERRED_LIGHTING_PASS,
```

Comment:
So deferred lighting is run before forward?

Reply:
Yes